src/coreclr/jit/valuenum.cpp

```diff
@@ -8139,11 +8139,19 @@ void Compiler::fgValueNumberTree(GenTree* tree)
             {
                 assert(!isVolatile); // We don't expect both volatile and invariant
 
-                // Is it a string literal? (it's always non-null)
-                if (addr->IsCnsIntOrI() && addr->IsIconHandle(GTF_ICON_STR_HDL))
+                // Is this invariant indirect expected to always return a non-null value?
+                if ((tree->gtFlags & GTF_IND_NONNULL))
                 {
-                    tree->gtVNPair = vnStore->VNPairForFunc(tree->TypeGet(), VNF_StrCns, addrNvnp);
-                    assert(addrXvnp.BothEqual() && (addrXvnp.GetLiberal() == ValueNumStore::VNForEmptyExcSet()));
+                    assert(tree->gtFlags & GTF_IND_NONFAULTING);
+                    tree->gtVNPair = vnStore->VNPairForFunc(tree->TypeGet(), VNF_NonNullIndirect, addrNvnp);
+                    if (addr->IsCnsIntOrI())
+                    {
+                        assert(addrXvnp.BothEqual() && (addrXvnp.GetLiberal() == ValueNumStore::VNForEmptyExcSet()));
+                    }
+                    else
```

Comment:
Can you describe when we'd hit this "else" clause?

Reply:
it's never hit but in theory should allow us to use that flag for let's say `IND(ADD(x, offset))` (I noticed some candidates in runtime getRuntimeLookupTree/fgExpandVirtualVtableCallTarget but didn't touch them yet. If you want - I can replace it with an assert instead until we need it.